modules/core/arrow-core-data/src/main/kotlin/arrow/core/Either.kt

```diff
@@ -6,11 +6,620 @@ import arrow.core.Either.Right
+ * functions can be thrown. Moreover, they may throw the same kind of exception
+ * (e.g. `IllegalArgumentException`) and thus it gets tricky tracking exactly where an exception came from.
+ *
+ * How then do we communicate an error? By making it explicit in the data type we return.
+ *
+ * ## Either vs Validated
+ *
+ * In general, `Validated` is used to accumulate errors, while `Either` is used to short-circuit a computation
+ * upon the first error. For more information, see the `Validated` vs `Either` section of the `Validated` documentation.
+ *
+ * By convention the right hand side of an `Either` is used to hold successful values.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ *
+ * val right: Either<String, Int> =
+ * //sampleStart
+ *  Either.Right(5)
+ * //sampleEnd
+ * fun main() {
+ *  println(right)
+ * }
+ * ```
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ *
+ * val left: Either<String, Int> =
+ * //sampleStart
+ *  Either.Left("Something went wrong")
+ * //sampleEnd
+ * fun main() {
+ *  println(left)
+ * }
+ * ```
+ * Because `Either` is right-biased, it is possible to define a Monad instance for it.
+ *
+ * Since we only ever want the computation to continue in the case of `Right` (as captured by the right-bias nature),
+ * we fix the left type parameter and leave the right one free.
+ *
+ * So the map and flatMap methods are right-biased:
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.flatMap
+ *
+ * //sampleStart
+ * val right: Either<String, Int> = Either.Right(5)
+ * val value = right.flatMap{ Either.Right(it + 1) }
+ * //sampleEnd
+ * fun main() {
+ *  println("value = $value")
+ * }
+ * ```
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.flatMap
+ *
+ * //sampleStart
+ * val left: Either<String, Int> = Either.Left("Something went wrong")
+ * val value = left.flatMap{ Either.Right(it + 1) }
+ * //sampleEnd
+ * fun main() {
+ *  println("value = $value")
+ * }
+ * ```
+ *
+ * ## Using Either instead of exceptions
+ *
+ * As a running example, we will have a series of functions that will:
+ *
+ * * Parse a string into an integer
+ * * Calculate the reciprocal
+ * * Convert the reciprocal into a string
+ *
+ * Using exception-throwing code, we could write something like this:
+ *
+ * ```kotlin:ank
+ * import arrow.core.Either
+ * import arrow.core.flatMap
+ *
+ * //sampleStart
+ * fun parse(s: String): Int =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt()
+ *   else throw NumberFormatException("$s is not a valid integer.")
+ *
+ * fun reciprocal(i: Int): Double =
+ *   if (i == 0) throw IllegalArgumentException("Cannot take reciprocal of 0.")
+ *   else 1.0 / i
+ *
+ * fun stringify(d: Double): String = d.toString()
+ * //sampleEnd
+ * ```
+ *
+ * Instead, let's make the fact that some of our functions can fail explicit in the return type.
+ *
```

Comment:
One section is missing here:
```suggestion
 * ```kotlin:ank
 * import arrow.core.Either
 * import arrow.core.flatMap
 * import arrow.core.left
 * import arrow.core.right
 *
 * //sampleStart
 * // Either Style
 * fun parse(s: String): Either<NumberFormatException, Int> =
 *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
 *   else Either.Left(NumberFormatException("$s is not a valid integer."))
 *
 * fun reciprocal(i: Int): Either<IllegalArgumentException, Double> =
 *   if (i == 0) Either.Left(IllegalArgumentException("Cannot take reciprocal of 0."))
 *   else Either.Right(1.0 / i)
 *
 * fun stringify(d: Double): String = d.toString()
 *
 * fun magic(s: String): Either<Exception, String> =
 *   parse(s).flatMap { reciprocal(it) }.map { stringify(it) }
 * //sampleEnd
 * ```
 * 
 * These calls to `parse` returns a `Left` and `Right` value
 *
```